Make initial nat gc async during Daemon initialization.

tommyp1ckles opened the pull request:
## Background

The initial CTMap GC has a hard coded timeout of 30 seconds: https://github.com/cilium/cilium/blob/4d016cc0c42de76f993877ea4dc3c618166ea62b/pkg/maps/ctmap/gc/gc.go#L227

This was designed to give more than sufficient time for this to complete. However, I would argue that such a timeout is not good practice as it's impossible to know all the possible environments this may be running on.  Specifically, a CPU constrained environment, such as a new Node starting possibly running other workloads, may experience unecessary crashes due to this if the ctmap + NAT map sweeps can't complete in time.  As well, this GC is also linked with the restored endpoint list - so a high number of restored endpoints also presents a factor in this timeout as well.

This PR seeks to make the initial CT GC async, such that newDaemon initialization continues without waiting for the first sweep to complete.  See [discussion](https://github.com/cilium/cilium/pull/34070#discussion_r1701046071) below for context regarding any possible issues resulting from implicit ordering in the newDaemon initialization.

Fixes: https://github.com/cilium/cilium/issues/32680

tommyp1ckles committed:
ctmap/gc: do not terminate agent fatal on ctmap gc init after timeout

If the agent is not completed its initial ctmap/nat GC scan within a
hardcoded 30 second timeout then the agent with terminate with
level=fatal.

This can cause issues in situations where the agent is running on an
environment with high resource contention.  This makes it possible for
the initial gc to timeout and put Cilium into a crash-loop.

The only affects the ctmap pressure metrics controller, which isn't
critical so instead of terminating the agent will now wait for the
initial scan async, and will emit a warning log after 30 seconds if
initial gc scan has not completed.

This will allow us to still catch issues in CI/environments where it
is unlikely that we should see these timeouts, while still avoiding
unnecessary termination.

Fixes: #32680

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ctmap/gc: add logging for ctmap initial gc durations.

With the previous commit the call to gc.Enable() ctmap gc no
longer waits for the initial pass to complete.

This section is tracked by a span in the daemon init but that will
no longer provide an accurate depiction of how long this gc scan
takes.

To help provide visibility here, we'll log the duration once the
initial scan has completed, as well as logging duration for NAT map
scan times to provide more a granular set of data on the relative gc
times.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

joestringer commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -220,15 +226,26 @@ func (gc *GC) Enable() {
+	// This is to notify users of potential issues affecting initial scan performance.
+	go func() {
+		select {
+		case <-initialScanComplete:
```
We'll have to be super careful around this. There's a decent chance that some other code in `daemon/cmd/daemon_main.go` relies on the behavior that when `Enable()` returns the connection tracker has been swept/garbage collected. Worst case this can have security implications.

Ideally this isn't the case, and if there _is_ any code that relies on the initial conntrack GC to complete, it should preferably depend on this same `<-initialScanComplete` channel before proceeding. I'm just not sure whether we're there yet.

tommyp1ckles replied:
@joestringer I've been trying to think of reasons why doing this async might cause issues.  I can't think of any but you may be right - we should err on the side of caution and simply bump the timeout as we intend to backport this.

tommyp1ckles replied:
I'm going to keep working on this branch, but as a simpler solution I'm going to open up this PR: https://github.com/cilium/cilium/pull/34136

joestringer replied:
Some background to my comment that we've faced some weird and interesting bugs in the past when changing the order of startup logic, because we didn't fully understand the ordering implications. To some degree that may just be the cost of progress, the code needs to be made more resilient and to more explicitly highlight the dependencies... we're just working through that tech debt now that there's a large init functions in the agent that rely on the order of operations within those functions. Things are definitely getting better with Hive/Cell init, but it's a process.

More specifically I recall that we have logic that does things like sweep the connection tracking table to ensure connections for stale Pods are cleared out. Let's say Cilium gets shut down. We have no way to tell how long Cilium was out, and some Pods get deleted/cleaned. Cilium starts back after some time. At some point it may be possible to reassign the same IP from that original Pod to a newly created Pod, so we want to ensure that the connection tracker doesn't contain entries from the old Pod. I think for this specific case there are some additional protections involving sweeping the CT table when provisioning new Pods, so we might not be entirely relying on that in this logic in this case. Other than that I know there's some fqdn/NameManager logic that interacts with CT state. There may be others, though they don't come to mind immediately.

To make this feedback more practical: As I understand this is being invoked from `startDaemon()`. Perhaps we can go through the rest of that function and at least look at the different other subsystems that are running logically after this function runs, and consider whether any of those subsystems may be relying on the status of the connection tracker. I'd be happy to brainstorm/discuss any potential links here or we can arrange something out-of-band and sync the state back to this thread.

tommyp1ckles replied:
> More specifically I recall that we have logic that does things like sweep the connection tracking table to ensure connections for stale Pods are cleared out.

Looking at this more deeply, the scans done by the `ctmap/gc.Enable()` do two things:

* Clean out expired connections (we see this [here](https://github.com/cilium/cilium/blob/main/pkg/maps/ctmap/gc/gc.go#L163) and [here](https://github.com/cilium/cilium/blob/main/pkg/maps/ctmap/gc/gc.go#L170) where we set the delete expired is always set to true)
* Internally (on the agent control-plane) [mark connections as zombies](https://github.com/cilium/cilium/blob/4e012f493ed7eda2a745be9dd22db5503ed06765/pkg/maps/ctmap/gc/gc.go#L141-L143) who's source IP belongs to a known endpoint to allow for working connections that may out live their FQDN DNS ttl times. 

On the other hand, endpoints do their own sweeps using `scrubIPsInConntrackTableLocked()` which will clear out any CTMap entries containing a src/dst address matching that of an endpoint: https://github.com/cilium/cilium/blob/main/pkg/endpoint/bpf.go#L1086-L1087.

> Cilium starts back after some time. At some point it may be possible to reassign the same IP from that original Pod to a newly created Pod, so we want to ensure that the connection tracker doesn't contain entries from the old Pod. I think for this specific case there are some additional protections involving sweeping the CT table when provisioning new Pods, 

So I think this portion is done outside of the regular ctmap GC "loop" if you're referring to the cleanup done by `ep.scrubIPsInConntrackTableLocked()`, this appears to be done on the first reconciliation of a particular endpoint.   I'll have to double check this but it doesn't sound like the ctmap GC needs to worry about this kind of protection as the only work it does is removing expired entries.

tommyp1ckles replied:
So if this assertion is correct (i.e. ctmap.GC only does expiry GC and refreshing zombie connections) then, disregarding the NAT GC for a moment, is there any other issue that might arise from ordering if we where to switch the ctmap GC to be async in the case of startDaemon.

One thing to maybe consider is any differences that async initial gc would cause wrt endpointmanager.  However, the [initial scan](https://github.com/cilium/cilium/blob/main/pkg/maps/ctmap/gc/gc.go#L163) does not do per endpoint GC (i.e. nil passed for ep in that case) so that should be safe.

joestringer replied:
Looking at the latest `git blame` for the conntrack GC lines in `startDaemon()` I found a commit that links to https://github.com/cilium/cilium/pull/32012 . There's some discussion there about CT GC timing, though no strong ordering expectations/recommendations.

Indeed you identified the FQDN subsystem, but in practice that subsystem seems to be starting up way earlier. The motivation there is to initialize the FQDN and DNS handling logic as early as possible to minimize impact on DNS traffic. I do wonder whether the FQDN NameManager GC should block on CT GC completing before starting so that it has the latest updates from the conntrack table, but currently that's also happening a lot earlier so I think the point is moot - whether or not we improve that is orthogonal to the goal of your PR.

I think that in practice the primary motivation for the CT GC logic is to make expiry timers effective, which is somewhat async anyway. The only thing I could logically think is if this is a new version of Cilium and it's trying to install a new datapath that somehow does something differently from the old datapath (such that it impacts CT entries => impacts GC). But even then to solve that theoretical problem the CT GC would have to happen _after_ initializing the new datapath, otherwise the new logic wouldn't be in place for all generated entries. So I think this is getting too abstract to be actionable.

You could throw a :fishing_pole_and_fish: towards @cilium/sig-datapath for review on this but after a bit of thinking through I don't know if there is something that this will break, and even if there is I'm not sure we'll be able to identify it. So maybe if we've thought a bit about this then we've done all that's necessary here to satisfy the concern and we should just move on towards proposing+merging+testing the implementation.

tommyp1ckles replied:
> Looking at the latest git blame for the conntrack GC lines in startDaemon() I found a commit that links to https://github.com/cilium/cilium/pull/32012 . There's some discussion there about CT GC timing, though no strong ordering expectations/recommendations.

This one makes sense, as the ct pass needs to potentially sweep endpoint maps for restored endpoints: https://github.com/cilium/cilium/blob/17efe6c460fa49938187629c8a17463e5a722fbc/pkg/maps/ctmap/gc/gc.go#L161

> The only thing I could logically think is if this is a new version of Cilium and it's trying to install a new datapath that somehow does something differently from the old datapath (such that it impacts CT entries => impacts GC). But even then to solve that theoretical problem the CT GC would have to happen after initializing the new datapath, otherwise the new logic wouldn't be in place for all generated entries. So I think this is getting too abstract to be actionable.

A practical example may be if someone switches between using jiffies and clock time, which I'm not entirely sure how that would be handled atm.

tommyp1ckles replied:
We can maybe quickly mention this in the community meeting discussion as well,  to see if anybody has any thoughts for why we might not want to do this.

tommyp1ckles replied:
Looks like you're the assignee for sig-datapath, and it won't let me add the group anymore, so community meeting discussion it is.

squeed replied:
w.r.t. pod lifecycle, we should never care about pods, only endpoints. Until a pod's endpoint gets a CNI DEL we **must** consider that endpoint as alive.

squeed replied:
as for FQDN, we now refuse to do any actual cleaning until GC has run at least 2 times, so we should be safe there. This has also been backported.

tommyp1ckles replied:
@squeed 

> w.r.t. pod lifecycle, we should never care about pods, only endpoints. Until a pod's endpoint gets a CNI DEL we must consider that endpoint as alive.

yeah, been sortof mixing terminology a bit here.  The main concern we had with endpoints was the per-endpoint ct cleaning, but that is run in a seperate lifecycle from the general ct/nat GC.

tommyp1ckles replied:
@joestringer re-requesting review, as @julianwiedmann mentioned lets take the opportunity to maybe merge this early in the development cycle.

tommyp1ckles committed:
ctmap: remove unused valid gc filter field.

While performing gc, this field would remove entries that had
a src/dst IP that wasn't part of a defined set of 'valid' IPs.

This filter was skipped if the ValidIPs field was left nil. There
appears to be no usage of this field.

Therefore, to simplify the GC logic this commit removes that
functionality.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ctmap/gc: use value semantics for GCFilter values.

Following simplification of the GCFilter filtering logic,
to help cleanup code, we can reuse common filters to make
code less verbose.

However, the GC filter does mutated in some cases to have it's
current time set for CT entry time expiry removal. To avoid issues
of shared pointers being mutated we can just make this pass by value
to avoid these concerns.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ctmap/gc: remove unused parameter/field for GC.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

joestringer reviewed (approved):
Ack per the last comment here https://github.com/cilium/cilium/pull/34070#discussion_r1716123440 .

christarazi reviewed (approved):
endpoint-specific changes makes sense to me.
